sdk/metrics/src/main/scala/org/typelevel/otel4s/sdk/metrics/InstrumentSelector.scala

```diff
@@ -0,0 +1,232 @@
+    obj match {
+      case other: InstrumentSelector =>
+        Hash[InstrumentSelector].eqv(this, other)
+      case _ =>
+        false
+    }
+}
+
+object InstrumentSelector {
+
+  /** A builder of an [[InstrumentSelector]].
+    */
+  sealed trait Builder {
+
+    /** Adds the given `tpe` as a predicate for an instrument.
+      *
+      * @param tpe
+      *   the instrument type to filter by
+      */
+    def withInstrumentType(tpe: InstrumentType): Builder
+
+    /** Adds the given `name` as a predicate for an instrument.
+      *
+      * @note
+      *   empty string is ignored
+      *
+      * @param name
+      *   the instrument name to filter by
+      */
+    def withInstrumentName(name: String): Builder
+
+    /** Adds the given `unit` as a predicate for an instrument.
+      *
+      * @note
+      *   empty string is ignored
+      *
+      * @param unit
+      *   the instrument type to filter by
+      */
+    def withInstrumentUnit(unit: String): Builder
+
+    /** Adds the given `name` as a predicate for a meter.
+      *
+      * @note
+      *   empty string is ignored
+      *
+      * @param name
+      *   the meter name to filter by
+      */
+    def withMeterName(name: String): Builder
+
+    /** Adds the given `version` as a predicate for a meter.
+      *
+      * @note
+      *   empty string is ignored
+      *
+      * @param version
+      *   the instrument type to filter by
+      */
+    def withMeterVersion(version: String): Builder
+
+    /** Adds the given `schemaUrl` as a predicate for a meter.
+      *
+      * @note
+      *   empty string is ignored
+      *
+      * @param schemaUrl
+      *   the instrument type to filter by
+      */
+    def withMeterSchemaUrl(schemaUrl: String): Builder
+
+    /** Creates an [[InstrumentSelector]] with the configuration of this
+      * builder.
+      */
+    def build: InstrumentSelector
+  }
+
+  /** Returns an empty [[Builder]] of an [[InstrumentSelector]].
+    */
+  def builder: Builder =
+    BuilderImpl()
+
+  implicit val instrumentSelectorHash: Hash[InstrumentSelector] =
+    Hash.by { selector =>
+      (
+        selector.instrumentType,
+        selector.instrumentName,
+        selector.instrumentUnit,
+        selector.meterName,
+        selector.meterVersion,
+        selector.meterSchemaUrl
+      )
+    }
+
+  implicit val instrumentSelectorShow: Show[InstrumentSelector] = {
+    Show.show { selector =>
+      def prop(focus: InstrumentSelector => Option[String], name: String) =
+        focus(selector).map(value => s"$name=$value")
+
+      Seq(
+        prop(_.instrumentType.map(_.toString), "instrumentType"),
+        prop(_.instrumentName, "instrumentName"),
+        prop(_.instrumentUnit, "instrumentUnit"),
+        prop(_.meterName, "meterName"),
+        prop(_.meterVersion, "meterVersion"),
+        prop(_.meterSchemaUrl, "meterSchemaUrl"),
+      ).collect { case Some(k) => k }.mkString("InstrumentSelector{", ", ", "}")
```

Comment:
We can use `StringJoiner` instead. https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/InstrumentSelector.java#L90-L111